internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,415 @@
+//
+// `name` specifies the name of the job object if a named job object is desired. If name
+// is an empty string, the job will not be assigned a name.
+//
+// `notifications` specifies if the job will be registered to receive notifications.
+// If this is false, `PollNotifications` will return immediately with error `errNotRegistered`.
+//
+// Returns a JobObject structure and an error if there is one.
+func Create(ctx context.Context, name string, notifications bool) (_ *JobObject, err error) {
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		if err != nil {
+			initIOErr = err
+			return
+		}
+		ioCompletionPort = h
+		go pollIOCP(ctx, h)
+	})
+
+	if initIOErr != nil && notifications {
+		return nil, initIOErr
+	}
+
+	var jobName *uint16
+	if name != "" {
+		jobName, err = windows.UTF16PtrFromString(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	jobHandle, err := windows.CreateJobObject(nil, jobName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			windows.Close(jobHandle)
+			delete(jobMap, uintptr(jobHandle))
+		}
+	}()
+
+	var mq *queue.MessageQueue
+	if notifications {
+		mq = queue.NewMessageQueue()
+		mapLock.Lock()
+		jobMap[uintptr(jobHandle)] = mq
+		mapLock.Unlock()
+		if err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+			return nil, err
+		}
+	}
+
+	return &JobObject{
+		handle: jobHandle,
+		mq:     mq,
+	}, nil
+}
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(limits *JobLimits) error {
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.MemoryLimitInBytes != 0 {
+		if err := job.SetMemoryLimit(limits.MemoryLimitInBytes); err != nil {
+			return errors.Wrap(err, "failed to set job object memory limit")
+		}
+	}
+
+	if limits.CPULimit != 0 {
+		if err := job.SetCPULimit(RateBased, limits.CPULimit); err != nil {
+			return errors.Wrap(err, "failed to set job object cpu limit")
+		}
+	} else if limits.CPUWeight != 0 {
+		if err := job.SetCPULimit(WeightBased, limits.CPUWeight); err != nil {
+			return errors.Wrap(err, "failed to set job object cpu limit")
+		}
+	}
+
+	if limits.MaxBandwidth != 0 || limits.MaxIOPS != 0 {
+		if err := job.SetIOLimit(limits.MaxBandwidth, limits.MaxIOPS); err != nil {
+			return errors.Wrap(err, "failed to set io limit on job object")
+		}
+	}
+	return nil
+}
+
+// SetCPULimit sets the CPU limit specified on the job object.
+func (job *JobObject) SetCPULimit(rateControlType CPURateControlType, rateControlValue uint32) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	var cpuInfo winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION
+	if rateControlType == WeightBased {
```

Comment:
kind of nitpicky, but I would prefer a switch statement here so we can error out nicely if not given a valid rate control type.